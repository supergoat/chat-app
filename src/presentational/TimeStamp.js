// @flow
import React, { Component } from 'react';
import { Text } from 'react-native';
import CAColors from "../common/CAColors";
import CAStyleSheet from "../common/CAStyleSheet";

/** ============================================================================
<TimeStamp
  date={Date}
  [style]={object}
/>
--------------------------------------------------------------------------------
Displays the date of the latest message
@param { string } date - The date of the latest message
@param { object } [style] - custom style
============================================================================= */
class TimeStamp extends Component<{
  date: Date,
  style?: { [name: string]: any }
}, {}> {

  /**
   * Returns:
   *  The time if the date is Today
   *  'YESTERDAY' if the date was yesterday
   *  The date if the date was before yesterday
   * @param { string } date - The date
   */
  getTimeStamp = (date: Date) => {
    const dateNow = new Date();
    const dayDifference = Math.abs(dateNow.getDate() - date.getDate());
    const monthDifference = Math.abs(dateNow.getMonth() - date.getMonth());
    const yearDifference = Math.abs(dateNow.getFullYear() - date.getFullYear());

    const isSameMonthAndYear =  monthDifference + yearDifference === 0;
    const today = isSameMonthAndYear && dayDifference === 0
    const yesterday = isSameMonthAndYear && dayDifference === 1

    if (today) {
      return this.formatTime(date)
    } else if (yesterday) {
      return 'YESTERDAY'
    } else {
      return this.formatDate(date);
    }
  }

  /**
   * Returns the time in the format HH:mm
   * @param { string } date - the date
   */
  formatTime = (date: Date) => {
    let hours = date.getHours();
    let minutes = date.getMinutes();

    if (hours < 10) hours = "0" + hours;
    if (minutes < 10) minutes = "0" + minutes;

    return `${hours}:${minutes}`
  }

  /**
   * Returns the date in the format of the time zone the user is in
   * For example
   *  UK: DD/MM/YY
   *  US: MM/DD/YY
   * @param { string } date - the date
   */
  formatDate = (date: Date) => {
    const options = {
      day: '2-digit',
      month: '2-digit',
      year: '2-digit'
    }
    return date.toLocaleDateString(undefined, options);
  }

  render() {
    const { date, style } = this.props;

    return (
      <Text style={[styles.timeStamp, style]}>
        {this.getTimeStamp(date)}
      </Text>
    )
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  timeStamp: {
    color: CAColors.oxfordBlue,
    fontSize: 12
  },
});


/* Export
============================================================================= */
export default TimeStamp;
