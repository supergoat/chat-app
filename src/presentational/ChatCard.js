// @flow
import React, { Component } from 'react';
import {
  Text,
  View,
  Image
} from 'react-native';
import ProfilePhoto from './ProfilePhoto';
import TimeStamp from './TimeStamp';
import NotificationBadge from './NotificationBadge';
import CAColors from '../common/CAColors';
import CAStyleSheet from '../common/CAStyleSheet';
import profilePhotoPlaceholder from '../assets/img/profile_photo_placeholder.png';

/* Constants
============================================================================= */
const CHAT_CARD_HEIGHT = 80;

/** ============================================================================
<ChatCard
  [profilePhoto]={string}
  name={string}
  latestMessage={{
    text: string,
    createdAt: Date
  }}
  unRead={number}
/>
--------------------------------------------------------------------------------
Displays information about a chat
@param { string } [profilePhoto] - The photo of the person the user is talking
@param { string } name - The name of the person the user is talking to
@param { object } latestMessage - The latest message of the chat
@param { string } unRead - The number of unRead messages
============================================================================= */
type Props = {
  profilePhoto?: string,
  name: string,
  latestMessage: {
    text: string,
    createdAt: Date
  },
  unRead: number
};
class ChatCard extends Component<Props> {
  static __cards__: Function;

  static defaultProps = {
    unRead: 0
  };

  render() {
    const { profilePhoto, name, latestMessage, unRead } = this.props;
    return (
      <View style={styles.chatCard}>
        <ProfilePhoto
          profilePhoto={profilePhoto}
          name={name}
          style={styles.profilePhoto}
        />

        <View style={styles.chatInfo}>
          <View style={styles.heading}>
            <Text style={styles.name}>{name}</Text>

            <TimeStamp
              date={latestMessage.createdAt}
              style={unRead > 0 ? styles.timeStamp : {}}
            />
          </View>

          <View style={styles.body}>
            <Text
              style={styles.latestMessage}
              ellipsizeMode="tail"
              numberOfLines={1}
            >
              {latestMessage.text}
            </Text>

            <NotificationBadge
              noOfNotifications={unRead}
              style={styles.notificationBadge}
            />
          </View>
        </View>
      </View>
    );
  }
}

/* StyleSheet
============================================================================= */
const styles = CAStyleSheet.create({
  chatCard: {
    height: CHAT_CARD_HEIGHT,
    backgroundColor: CAColors.white,
    alignItems: 'center',
    flexDirection: 'row'
  },
  profilePhoto: {
    marginLeft: 15
  },
  chatInfo: {
    height: CHAT_CARD_HEIGHT,
    flex: 1,
    margin: 10,
    borderBottomWidth: 1,
    borderColor: CAColors.gallery,
    justifyContent: 'center'
  },
  heading: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 5
  },
  body: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center'
  },
  name: {
    fontWeight: 'bold'
  },
  timeStamp: {
    color: CAColors.forestGreen
  },
  latestMessage: {
    color: CAColors.oxfordBlue,
    flex: 0.95
  }
});

/* Playground Cards - the different states of ChatCard
============================================================================= */
ChatCard.__cards__ = define => {
  let anHourAgo = new Date();
  anHourAgo.setHours(anHourAgo.getHours() - 1);

  let aDayAgo = new Date();
  aDayAgo.setDate(aDayAgo.getDate() - 1);

  let twoDaysAgo = new Date();
  twoDaysAgo.setDate(twoDaysAgo.getDate() - 2);

  define('Simple', () => {
    return (
      <ChatCard
        profilePhoto="https://placekitten.com/60/60"
        name="Alex Stone"
        latestMessage={{
          text: 'Hey whats up?',
          createdAt: aDayAgo
        }}
        unRead={10}
      />
    );
  });

  define('With notification over 99', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey man how is it going? Long time no see.',
          createdAt: anHourAgo
        }}
        unRead={100}
      />
    );
  });

  define('With truncated latest message text', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey man how is it going? Long time no see.',
          createdAt: aDayAgo
        }}
      />
    );
  });

  define('With the date as a timestamp and a notification', () => {
    return (
      <ChatCard
        name="Alex Stone"
        latestMessage={{
          text: 'Hey whats up?',
          createdAt: twoDaysAgo
        }}
        unRead={10}
      />
    );
  });
};

/* Export
============================================================================= */
export default ChatCard;
