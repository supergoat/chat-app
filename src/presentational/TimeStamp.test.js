// @flow
import React from 'react';
import TimeStamp from './TimeStamp';
import ChatCard from './ChatCard';
import renderer from 'react-test-renderer';

describe('TimeStamp', () => {
  // Custom constant date, so our snapshots don't fail when the date changes
  const customDate = new Date(2018, 1, 3, 9, 5);
  const RealDate = Date;

  beforeEach(() => {
    // Mock Date to always return our custom date
    global.Date = class extends RealDate {
      constructor() {
        return new RealDate(customDate);
      }
    };
  });

  afterEach(() => {
    global.Date = RealDate;
  });


  it('displays the time (HH:mm) when the date is today', () => {

    let anHourAgo = new Date();
    anHourAgo.setHours(anHourAgo.getHours() - 1);

    const tree = renderer
      .create(
        <TimeStamp
          date={anHourAgo}
        />)

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("displays 'YESTERDAY' when the date was yesterday", () => {
    let aDayAgo = new Date();
    aDayAgo.setDate(aDayAgo.getDate() - 1);

    const tree = renderer
      .create(
        <TimeStamp
          date={aDayAgo}
        />)

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('displays the date when the date was before yesterday', () => {
    let towDaysAgo = new Date();
    towDaysAgo.setDate(towDaysAgo.getDate() - 2);

    const tree = renderer
      .create(
        <TimeStamp
          date={towDaysAgo}
        />)

    expect(tree.toJSON()).toMatchSnapshot();
  });

});
