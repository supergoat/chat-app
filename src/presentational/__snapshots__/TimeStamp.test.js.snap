// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeStamp displays 'YESTERDAY' when the date was yesterday 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#333C4E",
        "fontSize": 12,
      },
      undefined,
    ]
  }
>
  YESTERDAY
</Text>
`;

exports[`TimeStamp displays the date when the date was before yesterday 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#333C4E",
        "fontSize": 12,
      },
      undefined,
    ]
  }
>
  18-02-01
</Text>
`;

exports[`TimeStamp displays the time (HH:mm) when the date is today 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#333C4E",
        "fontSize": 12,
      },
      undefined,
    ]
  }
>
  08:05
</Text>
`;
